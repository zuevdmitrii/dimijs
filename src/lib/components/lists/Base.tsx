import React, {useEffect, useMemo, useState} from 'react';
import {
  CrudErrorCodes,
  CrudEvents,
  ICrudFilter,
  ICrudList,
  ICrudPagination,
  ICrudSorting,
  ICrudStatus,
  ISource,
} from '../../sources/source';

interface IBaseProps<ItemType, KeyFieldType> {
  source: ISource<ItemType, KeyFieldType>;
  filter?: ICrudFilter<ItemType>[];
  pagination: ICrudPagination;
  sorting?: ICrudSorting<ItemType>[];
  RowTemplate: (props: {
    item: ItemType;
    source: ISource<ItemType, KeyFieldType>;
  }) => JSX.Element;
  PreloadTemplate?: () => JSX.Element;
  ErrorTemplate?: (props: {error: ICrudStatus}) => JSX.Element;
}
export const Base: <ItemType, KeyFieldType>(
  props: IBaseProps<ItemType, KeyFieldType>
) => JSX.Element = <ItemType, KeyFieldType>(
  props: IBaseProps<ItemType, KeyFieldType>
) => {
  const {
    source,
    pagination,
    filter,
    sorting,
    RowTemplate,
    ErrorTemplate,
    PreloadTemplate,
  } = props;
  const initData = useMemo(() => {
    return source.getSerializationData(filter || [], pagination, sorting);
  }, [source, filter, pagination, sorting]);
  const [list, setList] = useState<ItemType[]>(
    initData && (initData as ICrudList<ItemType>).data
      ? (initData as ICrudList<ItemType>).data
      : []
  );
  const [hasData, setHasData] = useState(
    !!(initData && (initData as ICrudList<ItemType>).data)
  );
  const [error, setError] = useState(
    initData && (initData as ICrudStatus).errorCode
      ? (initData as ICrudStatus)
      : {errorCode: CrudErrorCodes.OK}
  );
  const [isFirstRender, setIsFirstRender] = useState(true);
  useEffect(() => {
    const upList = async () => {
      const newList = await source.list(filter || [], pagination, sorting);
      if ((newList as ICrudStatus).errorCode) {
        setList([]);
        setError(newList as ICrudStatus);
      } else {
        setError({errorCode: CrudErrorCodes.OK});
        setList((newList as ICrudList<ItemType>).data);
        setHasData(true);
      }
    };
    const unsubscribes = [
      source.on(CrudEvents.onCreate, upList),
      source.on(CrudEvents.onDelete, upList),
      source.on(CrudEvents.onUpdate, upList),
    ];
    if (isFirstRender && !hasData) {
      setIsFirstRender(false);
      upList();
    }
    return () => void unsubscribes.forEach((f) => f());
  }, [source, filter, pagination, sorting, hasData, isFirstRender]);
  if (error.errorCode) {
    return <>{ErrorTemplate ? <ErrorTemplate error={error} /> : null}</>;
  }
  if (!hasData) {
    return <>{PreloadTemplate ? <PreloadTemplate /> : null}</>;
  }
  return (
    <>
      {list.map((row) => {
        return (
          <RowTemplate
            item={row}
            key={String(row[source.keyField])}
            source={source}
          />
        );
      })}
    </>
  );
};
