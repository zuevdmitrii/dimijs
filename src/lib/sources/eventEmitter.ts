type TCallback<Events> = (event: keyof Events, ...args: any) => void;

export class EventEmitter<Events> {
  private handlers: {[Property in keyof Events]?: TCallback<Events>[]} = {};
  on: (event: keyof Events, callback: TCallback<Events>) => () => void = (
    event,
    callback
  ) => {
    if (!this.handlers[event]) this.handlers[event] = [];
    this.handlers[event]?.push(callback);
    return () => {
      const index = this.handlers[event]?.indexOf(callback);
      if (index!==undefined && index>-1)
        this.handlers[event]?.splice(index, 1);
    };
  };
  fire: (event: keyof Events, ...args: any) => void = (event, ...args) => {
    if (this.handlers[event]) {
      this.handlers[event]?.forEach((f) => f(event, ...args));
    }
  };
}
