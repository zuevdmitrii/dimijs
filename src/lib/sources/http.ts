import {
  CrudErrorCodes,
  CrudEvents,
  ICrudFilter,
  ICrudList,
  ICrudPagination,
  ICrudSorting,
  ICrudStatus,
  Source,
} from './source';

export class Http<ItemType, KeyFieldType> extends Source<
  ItemType,
  KeyFieldType
> {
  // will be '${httpEndPoint}/{create|read|update|list}?'
  // create, delete and update = post methods
  // read, list = get
  public httpEndPoint: string = '/crud';
  create: (items: {
    [Property in keyof ItemType]+?: ItemType[Property];
  }[]) => Promise<ICrudStatus> = (items) => {
    return fetch(`${this.httpEndPoint}/create`, {
      method: 'POST',
      body: encodeURIComponent(JSON.stringify(items)),
    })
      .then((response) => {
        if (response.status === 200) return response.json();
        throw new Error(response.statusText);
      })
      .then((created) => {
        this.eventEmitter.fire(CrudEvents.onCreate, created);
        return {errorCode: CrudErrorCodes.OK};
      })
      .catch((reason) => {
        return {errorCode: CrudErrorCodes.ERROR, errorMessage: reason.message};
      });
  };
  read: (id: KeyFieldType) => Promise<ItemType | ICrudStatus | null> = (id) => {
    return fetch(
      `${this.httpEndPoint}/read?${encodeURIComponent(
        String(this.keyField)
      )}=${encodeURIComponent(String(id))}`,
      {
        method: 'GET',
      }
    )
      .then((response) => {
        if (response.status === 200) return response.json();
        throw new Error(response.statusText);
      })
      .catch((reason) => {
        return {errorCode: CrudErrorCodes.ERROR, errorMessage: reason.message};
      });
  };
  update: (data: {
    [Property in keyof ItemType]+?: ItemType[Property] | undefined;
  }) => Promise<ICrudStatus> = (data) => {
    return fetch(`${this.httpEndPoint}/update`, {
      method: 'POST',
      body: encodeURIComponent(JSON.stringify(data)),
    })
      .then((response) => {
        if (response.status === 200) return response.json();
        throw new Error(response.statusText);
      })
      .then((updated) => {
        this.eventEmitter.fire(CrudEvents.onUpdate, [updated]);
        return {errorCode: CrudErrorCodes.OK};
      })
      .catch((reason) => {
        return {errorCode: CrudErrorCodes.ERROR, errorMessage: reason.message};
      });
  };
  delete: (id: KeyFieldType) => Promise<ICrudStatus> = (id) => {
    return fetch(`${this.httpEndPoint}/delete`, {
      method: 'POST',
      body: encodeURIComponent(JSON.stringify({[this.keyField]: id})),
    })
      .then((response) => {
        if (response.status === 200) return response.json();
        throw new Error(response.statusText);
      })
      .then((deleted) => {
        if ((deleted as ICrudStatus).errorCode)
          return deleted;
        this.eventEmitter.fire(CrudEvents.onDelete, [deleted]);
        return {errorCode: CrudErrorCodes.OK};
      })
      .catch((reason) => {
        return {errorCode: CrudErrorCodes.ERROR, errorMessage: reason.message};
      });
  };
  list: (
    filter: ICrudFilter<ItemType>[],
    pagination: ICrudPagination,
    sorting?: ICrudSorting<ItemType>[]
  ) => Promise<ICrudStatus | ICrudList<ItemType>> = (
    filter,
    pagination,
    sorting
  ) => {
    return fetch(
      `${this.httpEndPoint}/list?params=${encodeURIComponent(
        JSON.stringify({filter, pagination, sorting})
      )}`,
      {
        method: 'GET',
      }
    )
      .then((response) => {
        if (response.status === 200) return response.json();
        throw new Error(response.statusText);
      })
      .then((jsonData) => {
        this.cache[JSON.stringify({filter, pagination, sorting})] = jsonData;
        return jsonData;
      })
      .catch((reason) => {
        return {errorCode: CrudErrorCodes.ERROR, errorMessage: reason.message};
      });
  };
}
