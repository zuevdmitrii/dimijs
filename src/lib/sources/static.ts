import {
  CrudAggregation,
  CrudErrorCodes,
  CrudEvents,
  CrudFilterFieldTypes,
  CrudFilterOperator,
  ICrudFilter,
  ICrudList,
  ICrudPagination,
  ICrudSorting,
  ICrudStatus,
  Source,
} from './source';

export class Static<ItemType, KeyFieldType>
  extends Source<ItemType, KeyFieldType>
{
  private data = [] as ItemType[];
  public delay: number = 0;
  constructor(
    keyField: keyof ItemType,
    deserializedData?: {
      data: ICrudList<ItemType> | ICrudStatus;
      filter?: ICrudFilter<ItemType>[];
      pagination?: ICrudPagination;
      sorting?: ICrudSorting<ItemType>[];
    }
  ) {
    super(keyField, deserializedData);
    if (
      deserializedData?.data &&
      (deserializedData?.data as ICrudList<ItemType>).data
    ) {
      this.data = (deserializedData.data as ICrudList<ItemType>).data.slice();
    } else {
      this.data = [];
    }
  }
  create: (items: ItemType[]) => Promise<ICrudStatus> = (items) => {
    return new Promise((resolve) => {
      setTimeout(() => {
        this.data = this.data.concat(items);
        this.eventEmitter.fire(CrudEvents.onCreate, items);
        resolve({
          errorCode: CrudErrorCodes.OK,
        });
      }, this.delay);
    });
  };
  read: (id: KeyFieldType) => Promise<ItemType | ICrudStatus | null> = (id) => {
    return new Promise((resolve) => {
      const row = this.data.find((item) => item[this.keyField] === id);
      resolve(row || null);
    });
  };
  update: (data: {
    [Property in keyof ItemType]+?: ItemType[Property] | undefined;
  }) => Promise<ICrudStatus> = (data) => {
    return new Promise((resolve) => {
      setTimeout(() => {
        const rowIndex = this.data.findIndex(
          (item) => item[this.keyField] === data[this.keyField]
        );
        if (rowIndex === -1) {
          resolve({
            errorCode: CrudErrorCodes.ERROR,
            errorMessage: `Not found by ${this.keyField.toString()}`,
          });
        } else {
          this.data[rowIndex] = {
            ...this.data[rowIndex],
            ...data,
          };
          this.eventEmitter.fire(CrudEvents.onUpdate, [this.data[rowIndex]]);
          resolve({errorCode: CrudErrorCodes.OK});
        }
      }, this.delay);
    });
  };
  delete: (id: KeyFieldType) => Promise<ICrudStatus> = (id) => {
    return new Promise((resolve) => {
      setTimeout(() => {
        const rowIndex = this.data.findIndex(
          (item) => item[this.keyField] === id
        );
        if (rowIndex === -1) {
          resolve({
            errorCode: CrudErrorCodes.ERROR,
            errorMessage: `Not found by ${this.keyField.toString()}`,
          });
        } else {
          const deleted = this.data.splice(rowIndex, 1);
          this.eventEmitter.fire(CrudEvents.onDelete, deleted);
          resolve({errorCode: CrudErrorCodes.OK});
        }
      }, this.delay);
    });
  };
  list: (
    filter: ICrudFilter<ItemType>[],
    pagination: ICrudPagination,
    sorting?: ICrudSorting<ItemType>[]
  ) => Promise<ICrudStatus | ICrudList<ItemType>> = (
    filter,
    pagination,
    sorting
  ) => {
    const generateFilterFunction: (
      filter: ICrudFilter<ItemType>[],
      agg: CrudAggregation
    ) => (item: ItemType) => boolean = (filter, agg) => {
      const conditions = [] as ((item: ItemType) => boolean)[];
      filter.forEach((p) => {
        if (p.type === CrudFilterFieldTypes.Aggregation) {
          conditions.push(generateFilterFunction(p.filter, p.aggregation));
        } else {
          conditions.push((item: ItemType) => {
            switch (p.operator) {
              case CrudFilterOperator.GT:
                return item[p.field] > p.value;
              case CrudFilterOperator.GTE:
                return item[p.field] >= p.value;
              case CrudFilterOperator.LT:
                return item[p.field] < p.value;
              case CrudFilterOperator.LTE:
                return item[p.field] <= p.value;
              case CrudFilterOperator.NOTEQ:
                return item[p.field] !== p.value;
              case CrudFilterOperator.LIKE: {
                if (typeof item[p.field] === 'string')
                  return (item[p.field] as string).indexOf(`${p.value}`) > -1;
                return false;
              }
              case CrudFilterOperator.EQ:
              default:
                return item[p.field] === p.value;
            }
          });
        }
      });
      return (item) => {
        if (agg === CrudAggregation.AND)
          return conditions.every((fn) => fn(item));
        return conditions.some((fn) => fn(item));
      };
    };
    const filterFunction = generateFilterFunction(filter, CrudAggregation.AND);
    return new Promise((resolve) => {
      const filtered = this.data.filter(filterFunction);
      if (sorting) {
        sorting.forEach((sortField) => {
          filtered.sort((itemA, itemB) => {
            const isNum =
              typeof itemA[sortField.field] === 'number' ||
              typeof itemB[sortField.field] === 'number';
            if (isNum) {
              return sortField.direction === 'DESC'
                ? (itemB[sortField.field] as number) -
                    (itemA[sortField.field] as number)
                : (itemA[sortField.field] as number) -
                    (itemB[sortField.field] as number);
            }
            return sortField.direction === 'DESC'
              ? (itemB[sortField.field] as string).localeCompare(
                  itemA[sortField.field] as string
                )
              : (itemA[sortField.field] as string).localeCompare(
                  itemB[sortField.field] as string
                );
          });
        });
      }
      const startIndex = pagination.page * pagination.countOnPage;
      const result = filtered.slice(
        startIndex,
        startIndex + pagination.countOnPage
      );
      setTimeout(() => {
        const forReturn = {
          data: result,
          meta: {
            hasNextPage: !!filtered[startIndex + pagination.countOnPage],
          },
        };
        this.cache[JSON.stringify({filter, pagination, sorting})] = forReturn;
        resolve(forReturn);
      }, this.delay);
    });
  };
}
