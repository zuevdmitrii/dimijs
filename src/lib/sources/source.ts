import {EventEmitter} from './eventEmitter';

export enum CrudErrorCodes {
  OK = 0,
  ERROR = 1,
}
export interface ICrudStatus {
  errorCode: CrudErrorCodes;
  errorMessage?: string;
}
export enum CrudFilterOperator {
  EQ,
  GT,
  GTE,
  LT,
  LTE,
  NOTEQ,
  LIKE,
}
export enum CrudAggregation {
  AND,
  OR,
}
export enum CrudFilterFieldTypes {
  Aggregation,
  Field,
}
export type ICrudFilterField<ItemType, T extends keyof ItemType> = {
  operator?: CrudFilterOperator;
  field: T;
  value: ItemType[T];
  type: CrudFilterFieldTypes.Field;
};

export interface ICrudAggregationFilterField<ItemType> {
  type: CrudFilterFieldTypes.Aggregation;
  aggregation: CrudAggregation;
  filter: ICrudFilter<ItemType>[];
}
export type ICrudFilter<ItemType> =
  | ICrudAggregationFilterField<ItemType>
  | ICrudFilterField<ItemType, keyof ItemType>;

export interface ICrudPagination {
  page: number;
  countOnPage: number;
}
export interface ICrudSorting<ItemType> {
  field: keyof ItemType;
  direction?: 'DESC' | 'ASC';
}
export interface ICrudList<ItemType> {
  data: ItemType[];
  meta?: {
    hasNextPage?: boolean;
  };
}
export enum CrudEvents {
  onCreate = 'onCreate',
  onDelete = 'onDelete',
  onUpdate = 'onUpdate',
}

export interface ISource<ItemType, KeyFieldType> {
  keyField: keyof ItemType;
  create: (items: ItemType[]) => Promise<ICrudStatus>;
  read: (id: KeyFieldType) => Promise<ItemType | ICrudStatus | null>;
  update: (data: {
    [Property in keyof ItemType]+?: ItemType[Property];
  }) => Promise<ICrudStatus>;
  delete: (id: KeyFieldType) => Promise<ICrudStatus>;
  list: (
    filter: ICrudFilter<ItemType>[],
    pagination: ICrudPagination,
    sorting?: ICrudSorting<ItemType>[]
  ) => Promise<ICrudList<ItemType> | ICrudStatus>;
  // return data from local cache syncroniously for using in component render
  getSerializationData: (
    filter: ICrudFilter<ItemType>[],
    pagination: ICrudPagination,
    sorting?: ICrudSorting<ItemType>[]
  ) => ICrudList<ItemType> | ICrudStatus | null;
  on: (
    event: CrudEvents,
    callback: (event: CrudEvents, data: ItemType[]) => void
  ) => () => void;
}

export class Source<ItemType, KeyFieldType>
  implements ISource<ItemType, KeyFieldType>
{
  protected cache = {} as {
    [propname: string]: ICrudList<ItemType> | ICrudStatus;
  };
  protected eventEmitter = new EventEmitter<{
    [Property in CrudEvents]: Property;
  }>();

  public keyField: keyof ItemType;
  constructor(
    keyField: keyof ItemType,
    deserializedData?: {
      data: ICrudList<ItemType> | ICrudStatus;
      filter?: ICrudFilter<ItemType>[];
      pagination?: ICrudPagination;
      sorting?: ICrudSorting<ItemType>[];
    }
  ) {
    this.keyField = keyField;
    if (
      deserializedData &&
      deserializedData.pagination &&
      deserializedData.filter
    ) {
      const {filter, pagination, sorting} = deserializedData;
      this.cache[JSON.stringify({filter, pagination, sorting})] =
        deserializedData.data;
    }
  }
  getSerializationData: (
    filter: ICrudFilter<ItemType>[],
    pagination: ICrudPagination,
    sorting?: ICrudSorting<ItemType>[]
  ) => ICrudList<ItemType> | ICrudStatus | null = (
    filter,
    pagination,
    sorting
  ) => {
    return this.cache[JSON.stringify({filter, pagination, sorting})] || null;
  };
  on: (
    event: CrudEvents,
    callback: (event: CrudEvents, data: ItemType[]) => void
  ) => () => void = (event, callback) => {
    return this.eventEmitter.on(event, callback);
  };
  create: (items: ItemType[]) => Promise<ICrudStatus> = () => {
    throw new Error('Not implemeted');
  };
  read: (id: KeyFieldType) => Promise<ItemType | ICrudStatus | null> = () => {
    throw new Error('Not implemeted');
  };
  update: (data: {
    [Property in keyof ItemType]+?: ItemType[Property] | undefined;
  }) => Promise<ICrudStatus> = () => {
    throw new Error('Not implemeted');
  };
  delete: (id: KeyFieldType) => Promise<ICrudStatus> = () => {
    throw new Error('Not implemeted');
  };
  list: (
    filter: ICrudFilter<ItemType>[],
    pagination: ICrudPagination,
    sorting?: ICrudSorting<ItemType>[]
  ) => Promise<ICrudStatus | ICrudList<ItemType>> = () => {
    throw new Error('Not implemeted');
  };
}
