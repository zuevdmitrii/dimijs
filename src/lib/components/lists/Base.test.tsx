import React from 'react';
import {act, render, screen, waitFor} from '@testing-library/react';
import {Base} from './Base';
import {Static} from '../../sources/static';
import {CrudErrorCodes} from '../../sources/source';
import {resolve} from 'node:path/win32';

interface IDemoData {
  id: number;
  value: string;
}
test('render empty list', async () => {
  const source = new Static<IDemoData, number>('id');
  const pagination = {page: 0, countOnPage: 10};
  const {container} = render(
    <div>
      <Base
        source={source}
        pagination={pagination}
        RowTemplate={({item}) => {
          return <span>{item.id}</span>;
        }}
        PreloadTemplate={() => <span>Loading...</span>}
      />
    </div>
  );
  expect(container.firstChild).toContainHTML('Loading');
  await act(async () => {
    await source.create([{id: 1, value: 'test'}]);
  });
  await waitFor(() => screen.getByText('1'));
  expect(container.firstChild).toContainHTML('1');
  const forSerialization = source.getSerializationData([], pagination);
  expect(forSerialization).toMatchSnapshot();
});

test('render empty list with empty preload', async () => {
  const source = new Static<IDemoData, number>('id');
  const pagination = {page: 0, countOnPage: 10};
  const {container} = render(
    <div>
      <Base
        source={source}
        pagination={pagination}
        RowTemplate={({item}) => {
          return <span>{item.id}</span>;
        }}
      />
    </div>
  );
  expect(container.firstChild).toBeEmptyDOMElement();
});

test('render deserialized list', async () => {
  const serverData = {
    data: [{id: 1, value: 'test'}],
    meta: {hasNextPage: false},
  };
  const source = new Static<IDemoData, number>('id', {
    data: serverData,
    filter: [],
    pagination: {page: 0, countOnPage: 10},
  });
  const pagination = {page: 0, countOnPage: 10};
  const {container} = render(
    <div>
      <Base
        source={source}
        pagination={pagination}
        RowTemplate={({item}) => {
          return <span>{item.id}</span>;
        }}
        PreloadTemplate={() => <span>Loading...</span>}
      />
    </div>
  );
  expect(container.firstChild).toContainHTML('1');
  const forSerialization = source.getSerializationData([], pagination);
  expect(forSerialization).toMatchObject(serverData);
});

test('render error template', async () => {
  const serverData = {
    errorCode: CrudErrorCodes.ERROR,
    errorMessage: 'Demo error',
  };
  const source = new Static<IDemoData, number>('id', {
    data: serverData,
    filter: [],
    pagination: {page: 0, countOnPage: 10},
  });
  const pagination = {page: 0, countOnPage: 10};
  const {container} = render(
    <div>
      <Base
        source={source}
        pagination={pagination}
        RowTemplate={({item}) => {
          return <span>{item.id}</span>;
        }}
        ErrorTemplate={({error}) => <span>Err: {error.errorMessage}...</span>}
      />
    </div>
  );
  expect(container.firstChild).toContainHTML('Err: Demo error...');
  const forSerialization = source.getSerializationData([], pagination);
  expect(forSerialization).toMatchObject(serverData);
});

test('render empty error template', async () => {
  const serverData = {
    errorCode: CrudErrorCodes.ERROR,
    errorMessage: 'Demo error',
  };
  const source = new Static<IDemoData, number>('id', {
    data: serverData,
    filter: [],
    pagination: {page: 0, countOnPage: 10},
  });
  const pagination = {page: 0, countOnPage: 10};
  const {container} = render(
    <div>
      <Base
        source={source}
        pagination={pagination}
        RowTemplate={({item}) => {
          return <span>{item.id}</span>;
        }}
      />
    </div>
  );
  expect(container.firstChild).toBeEmptyDOMElement();
});

test('render error tmpl after list method', async () => {
  const source = new Static<IDemoData, number>('id');
  const pagination = {page: 0, countOnPage: 10};
  const {container} = render(
    <div>
      <Base
        source={source}
        pagination={pagination}
        RowTemplate={({item}) => {
          return <span>{item.id}</span>;
        }}
        PreloadTemplate={() => <span>Loading...</span>}
        ErrorTemplate={({error}) => <span>Err: {error.errorMessage}...</span>}
      />
    </div>
  );
  // @ts-ignore
  source.list = () => {
    return new Promise((resolve) => {
      resolve({errorCode: CrudErrorCodes.ERROR, errorMessage: 'Runtime error'});
    });
  };
  expect(container.firstChild).toContainHTML('Loading');
  await act(async () => {
    await source.create([{id: 1, value: 'test'}]);
  });
  await waitFor(() => screen.getByText('Err: Runtime error...'));
  expect(container.firstChild).toContainHTML('Err: Runtime error...');
});

test('render empty error tmpl after list method', async () => {
  const source = new Static<IDemoData, number>('id');
  const pagination = {page: 0, countOnPage: 10};
  const {container} = render(
    <div>
      <Base
        source={source}
        pagination={pagination}
        RowTemplate={({item}) => {
          return <span>{item.id}</span>;
        }}
        PreloadTemplate={() => <span>Loading...</span>}
        ErrorTemplate={({error}) => (
          <span>Err: {error.errorMessage || 'unknown'}...</span>
        )}
      />
    </div>
  );
  // @ts-ignore
  source.list = () => {
    return new Promise((resolve) => {
      resolve({errorCode: CrudErrorCodes.ERROR, errorMessage: ''});
    });
  };
  expect(container.firstChild).toContainHTML('Loading');
  await act(async () => {
    await source.create([{id: 1, value: 'test'}]);
  });
  await waitFor(() => screen.getByText('Err: unknown...'));
  expect(container.firstChild).toContainHTML('Err: unknown...');
});

test('render deserialized list after first request', async () => {
  const serverData = {
    data: [{id: 1, value: 'test'}],
    meta: {hasNextPage: false},
  };
  const source = new Static<IDemoData, number>('id', {
    data: serverData,
  });
  const pagination = {page: 0, countOnPage: 10};
  const {container} = render(
    <div>
      <Base
        source={source}
        pagination={pagination}
        RowTemplate={({item}) => {
          return <span>{item.id}</span>;
        }}
        PreloadTemplate={() => <span>Loading...</span>}
      />
    </div>
  );
  expect(container.firstChild).toContainHTML('Loading');
  await waitFor(() => screen.getByText('1'));
  expect(container.firstChild).toContainHTML('1');
});