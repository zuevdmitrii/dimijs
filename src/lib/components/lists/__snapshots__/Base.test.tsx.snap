// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render empty list 1`] = `
Object {
  "data": Array [
    Object {
      "id": 1,
      "value": "test",
    },
  ],
  "meta": Object {
    "hasNextPage": false,
  },
}
`;
